File: frontend/src/services/api/preferences.ts
/**
 * User Preferences API Service
 *
 * Handles all user preferences CRUD operations including:
 * - Fiscal year configuration
 * - Essential vs discretionary categories
 * - Investment account mappings
 * - Income source categories
 * - Budget defaults
 * - Display/format preferences
 * - Anomaly detection settings
 * - Recurring transaction settings
 */

import { apiClient } from './client'
import type { SalaryComponents, RsuGrant, GrowthAssumptions } from '@/types/salary'

// Types
export interface UserPreferences {
  id: number

  // 1. Fiscal Year
  fiscal_year_start_month: number

  // 2. Essential Categories
  essential_categories: string[]

  // 3. Investment Mappings
  investment_account_mappings: Record<string, string>

  // 4. Income Classification (by tax treatment)
  taxable_income_categories: string[]
  investment_returns_categories: string[]
  non_taxable_income_categories: string[]
  other_income_categories: string[]

  // 5. Budget Defaults
  default_budget_alert_threshold: number
  auto_create_budgets: boolean
  budget_rollover_enabled: boolean

  // 6. Display Preferences
  number_format: 'indian' | 'international'
  currency_symbol: string
  currency_symbol_position: 'before' | 'after'
  default_time_range: string
  display_currency: string

  // 7. Anomaly Settings
  anomaly_expense_threshold: number
  anomaly_types_enabled: string[]
  auto_dismiss_recurring_anomalies: boolean

  // 8. Recurring Settings
  recurring_min_confidence: number
  recurring_auto_confirm_occurrences: number

  // 9. Spending Rule Targets
  needs_target_percent: number
  wants_target_percent: number
  savings_target_percent: number

  // 10. Credit Card Limits
  credit_card_limits: Record<string, number>

  // 11. Earning Start Date
  earning_start_date: string | null
  use_earning_start_date: boolean

  // 12. Fixed/Mandatory Monthly Expenses
  fixed_expense_categories: string[] | string

  // 13. Savings & Investment Targets
  savings_goal_percent: number
  monthly_investment_target: number

  // 14. Payday Configuration
  payday: number

  // 15. Tax Regime Preference
  preferred_tax_regime: string

  // 16. Excluded Accounts
  excluded_accounts: string[] | string

  // 17. Notification Preferences
  notify_budget_alerts: boolean
  notify_anomalies: boolean
  notify_upcoming_bills: boolean
  notify_days_ahead: number

  // 18. Tax display
  show_tds_schedule: boolean

  // 19. EPF withdrawal taxability
  epf_withdrawal_taxable: boolean
  epf_taxable_percent: number

  // 20. Salary TDS treatment: are recorded salary amounts net of TDS (default)
  // or gross (pre-tax)?
  salary_is_net_of_tds: boolean

  // Salary & Tax Projections
  salary_structure: Record<string, SalaryComponents>
  rsu_grants: RsuGrant[]
  growth_assumptions: GrowthAssumptions

  // AI Assistant
  ai_provider: string | null
  ai_model: string | null

  // Metadata
  created_at: string | null
  updated_at: string | null
}

// Partial update type
export type UserPreferencesUpdate = Partial<Omit<UserPreferences, 'id' | 'created_at' | 'updated_at'>>

// Section-specific update types
export interface FiscalYearConfig {
  fiscal_year_start_month: number
}

export interface EssentialCategoriesConfig {
  essential_categories: string[]
}

export interface InvestmentMappingsConfig {
  investment_account_mappings: Record<string, string>
}

export interface IncomeSourcesConfig {
  taxable_income_categories: string[]
  investment_returns_categories: string[]
  non_taxable_income_categories: string[]
  other_income_categories: string[]
}

export interface BudgetDefaultsConfig {
  default_budget_alert_threshold: number
  auto_create_budgets: boolean
  budget_rollover_enabled: boolean
}

export interface DisplayPreferencesConfig {
  number_format: 'indian' | 'international'
  currency_symbol: string
  currency_symbol_position: 'before' | 'after'
  default_time_range: string
  display_currency: string
}

export interface AnomalySettingsConfig {
  anomaly_expense_threshold: number
  anomaly_types_enabled: string[]
  auto_dismiss_recurring_anomalies: boolean
}

export interface RecurringSettingsConfig {
  recurring_min_confidence: number
  recurring_auto_confirm_occurrences: number
}

export interface SpendingRuleConfig {
  needs_target_percent: number
  wants_target_percent: number
  savings_target_percent: number
}

export interface CreditCardLimitsConfig {
  credit_card_limits: Record<string, number>
}

export interface EarningStartDateConfig {
  earning_start_date: string | null
  use_earning_start_date: boolean
}

export interface SalaryStructureConfig {
  salary_structure: Record<string, SalaryComponents>
}

export interface RsuGrantsConfig {
  rsu_grants: RsuGrant[]
}

export interface GrowthAssumptionsConfig {
  growth_assumptions: GrowthAssumptions
}

/**
 * `/api/exchange-rates`.
 *
 * Three response variants from one handler, distinguished by these flags:
 *  - live fetch: `fetched_at` set, no flags
 *  - stale cache (upstream fetch failed): `stale: true`, `fetched_at` from the
 *    last successful fetch
 *  - hardcoded fallback: `fallback: true`, `fetched_at: null`, and
 *    `fallback_as_of` carrying the date the baked-in table was captured
 *
 * `fallback_as_of` was missing from this interface, so nothing read it and
 * `useExchangeRate` stamped fallback rates with `new Date()` -- presenting a
 * table that could be months old as "fetched just now". Verified against the
 * handler on 2026-07-27.
 */
export interface ExchangeRatesResponse {
  base: string
  rates: Record<string, number>
  fetched_at?: number | null
  stale?: boolean
  fallback?: boolean
  /** ISO date (YYYY-MM-DD) the hardcoded table was captured. Only on fallback. */
  fallback_as_of?: string
  /** True when the response is a dated historical rate, not the latest. */
  historical?: boolean
  /**
   * ISO date the rate was actually published. Precedes the requested date
   * across a weekend or holiday. Only on historical responses.
   */
  as_of?: string
  /** The date that was asked for. Only on historical responses. */
  requested_date?: string
}

// Helper to create section-specific updaters
function createSectionUpdater<T>(endpoint: string) {
  return async (config: T): Promise<UserPreferences> => {
    const response = await apiClient.put<UserPreferences>(`/api/preferences/${endpoint}`, config)
    return response.data
  }
}

// Service
export const preferencesService = {
  /**
   * Get current user preferences
   */
  async getPreferences(): Promise<UserPreferences> {
    const response = await apiClient.get<UserPreferences>('/api/preferences')
    return response.data
  },

  /**
   * Update user preferences (partial update supported)
   */
  async updatePreferences(updates: UserPreferencesUpdate): Promise<UserPreferences> {
    const response = await apiClient.put<UserPreferences>('/api/preferences', updates)
    return response.data
  },

  /**
   * Reset all preferences to defaults
   */
  async resetPreferences(): Promise<UserPreferences> {
    const response = await apiClient.post<UserPreferences>('/api/preferences/reset')
    return response.data
  },

  // Section-specific endpoints for granular updates
  updateFiscalYear: createSectionUpdater<FiscalYearConfig>('fiscal-year'),
  updateEssentialCategories: createSectionUpdater<EssentialCategoriesConfig>('essential-categories'),
  updateInvestmentMappings: createSectionUpdater<InvestmentMappingsConfig>('investment-mappings'),
  updateIncomeSources: createSectionUpdater<IncomeSourcesConfig>('income-sources'),
  updateBudgetDefaults: createSectionUpdater<BudgetDefaultsConfig>('budget-defaults'),
  updateDisplayPreferences: createSectionUpdater<DisplayPreferencesConfig>('display'),
  updateAnomalySettings: createSectionUpdater<AnomalySettingsConfig>('anomaly-settings'),
  updateRecurringSettings: createSectionUpdater<RecurringSettingsConfig>('recurring-settings'),
  updateSpendingRule: createSectionUpdater<SpendingRuleConfig>('spending-rule'),
  updateCreditCardLimits: createSectionUpdater<CreditCardLimitsConfig>('credit-card-limits'),
  updateEarningStartDate: createSectionUpdater<EarningStartDateConfig>('earning-start-date'),
  updateSalaryStructure: createSectionUpdater<SalaryStructureConfig>('salary-structure'),
  updateRsuGrants: createSectionUpdater<RsuGrantsConfig>('rsu-grants'),
  updateGrowthAssumptions: createSectionUpdater<GrowthAssumptionsConfig>('growth-assumptions'),

  async getStockPrice(
    symbol: string,
    onDate?: string,
  ): Promise<{ symbol: string; price: number; currency: string; as_of: string | null }> {
    const response = await apiClient.get<{
      symbol: string
      price: number
      currency: string
      as_of: string | null
    }>(`/api/stock-price/${encodeURIComponent(symbol)}`, {
      params: onDate ? { on_date: onDate } : undefined,
    })
    return response.data
  },

  /**
   * Exchange rates for `base`, latest or as published on `onDate`.
   *
   * `onDate` exists so a historical value is converted at the FX rate that
   * applied then. Converting an RSU vest-date stock price at today's rate mixed
   * vintages and overstated the vested line by the FX drift since (measured 9%
   * on a 2025-08-15 AMZN vest: USD/INR 87.46 then vs 95.34 on 2026-08-03).
   */
  async getExchangeRates(
    base: string = 'INR',
    onDate?: string,
  ): Promise<ExchangeRatesResponse> {
    // Generic on `get`, not just on the return type: without it `response.data`
    // is `any`, so the declared shape was an unchecked assertion and a backend
    // rename would have surfaced as `undefined` at the render site instead of
    // an error here. Shape confirmed live on 2026-07-27 (29 currencies).
    const response = await apiClient.get<ExchangeRatesResponse>('/api/exchange-rates', {
      params: onDate ? { base, on_date: onDate } : { base },
    })
    return response.data
  },
}

File: backend/src/ledger_sync/api/preferences.py
"""User Preferences API endpoints (general + sub-section setters).

AI-config endpoints live in api/preferences_ai.py (also mounted on the
same router via include_router below).
"""

from __future__ import annotations

import json
from datetime import UTC, datetime

from fastapi import APIRouter

from ledger_sync.api.deps import CurrentUser, DatabaseSession
from ledger_sync.api.preferences_ai import router as ai_router
from ledger_sync.api.preferences_helpers import (
    AnomalySettingsConfig,
    BudgetDefaultsConfig,
    CapitalLossConfig,
    CreditCardLimitsConfig,
    DisplayPreferencesConfig,
    EarningStartDateConfig,
    EssentialCategoriesConfig,
    FiscalYearConfig,
    IncomeSourcesConfig,
    InvestmentMappingsConfig,
    RecurringSettingsConfig,
    SpendingRuleConfig,
    UserPreferencesResponse,
    UserPreferencesUpdate,
    _get_or_create_preferences,
    _model_to_response,
    _update_section,
)
from ledger_sync.schemas.salary import (
    GrowthAssumptionsConfig,
    RsuGrantsConfig,
    SalaryStructureConfig,
)

router = APIRouter(prefix="/api/preferences", tags=["preferences"])
router.include_router(ai_router)


# ----- API Endpoints -----


@router.get("")
def get_preferences(
    current_user: CurrentUser,
    session: DatabaseSession,
) -> UserPreferencesResponse:
    """Get current user preferences."""
    prefs = _get_or_create_preferences(session, current_user)
    return _model_to_response(prefs)


@router.put("")
def update_preferences(
    current_user: CurrentUser,
    updates: UserPreferencesUpdate,
    session: DatabaseSession,
) -> UserPreferencesResponse:
    """Update user preferences (partial update supported)."""
    prefs = _get_or_create_preferences(session, current_user)

    # Apply updates for non-None fields
    update_data = updates.model_dump(exclude_none=True)

    for field, value in update_data.items():
        # Convert lists/dicts to JSON strings for storage
        if isinstance(value, (list, dict)):
            value = json.dumps(value)
        setattr(prefs, field, value)

    prefs.updated_at = datetime.now(UTC)
    session.commit()
    session.refresh(prefs)

    return _model_to_response(prefs)


@router.post("/reset")
def reset_preferences(
    current_user: CurrentUser,
    session: DatabaseSession,
) -> UserPreferencesResponse:
    """Reset all preferences to defaults (empty values for data-dependent fields)."""
    prefs = _get_or_create_preferences(session, current_user)

    # Reset to defaults - data-dependent fields start empty
    prefs.fiscal_year_start_month = 4
    prefs.essential_categories = json.dumps([])  # Empty - user configures after upload
    prefs.investment_account_mappings = json.dumps({})  # Empty - user configures after upload
    # All FOUR income lists are emptied together, and that symmetry is
    # load-bearing rather than incidental. They are a PARTITION written by one
    # exclusive-assignment UI, so every reader decides the empty-means-unset
    # fallback for the GROUP, not per field: core/analytics/base.py::
    # _any_income_list_configured and store/preferencesStore.ts::
    # withIncomeClassificationDefaults both fall back to the shipped defaults
    # only while NO list holds a user choice. Seeding one list here and leaving
    # its three siblings "[]" makes the group look configured, which strands
    # taxable/investment/other permanently empty -- on a real ledger that books
    # every salary row as unclassified income and reports zero taxable income on
    # the Tax Planning page. It also trips the Settings auto-classify recovery,
    # whose guard is an OR across all four lists.
    prefs.taxable_income_categories = json.dumps([])
    prefs.investment_returns_categories = json.dumps([])
    prefs.non_taxable_income_categories = json.dumps([])
    prefs.other_income_categories = json.dumps([])
    # Emptied like every other data-dependent field. Unlike the four income
    # lists above there is no shipped default to fall back to: an empty list
    # means "classify nothing", so a reset leaves realised losses counted as
    # expenses exactly as they were before this preference existed.
    prefs.capital_loss_categories = json.dumps([])
    prefs.default_budget_alert_threshold = 80.0
    prefs.auto_create_budgets = False
    prefs.budget_rollover_enabled = False
    prefs.number_format = "indian"
    prefs.currency_symbol = "₹"
    prefs.currency_symbol_position = "before"
    prefs.default_time_range = "all_time"
    prefs.display_currency = "INR"
    prefs.anomaly_expense_threshold = 2.0
    prefs.anomaly_types_enabled = json.dumps(
        ["high_expense", "unusual_category", "large_transfer", "budget_exceeded"],
    )
    prefs.auto_dismiss_recurring_anomalies = True
    prefs.recurring_min_confidence = 50.0
    prefs.recurring_auto_confirm_occurrences = 6
    prefs.needs_target_percent = 50.0
    prefs.wants_target_percent = 30.0
    prefs.savings_target_percent = 20.0
    prefs.credit_card_limits = json.dumps({})
    prefs.earning_start_date = None
    prefs.use_earning_start_date = False
    prefs.fixed_expense_categories = json.dumps([])
    prefs.savings_goal_percent = 20.0
    prefs.monthly_investment_target = 0.0
    prefs.payday = 1
    prefs.preferred_tax_regime = "new"
    prefs.excluded_accounts = json.dumps([])
    prefs.notify_budget_alerts = True
    prefs.notify_anomalies = True
    prefs.notify_upcoming_bills = True
    prefs.notify_days_ahead = 7
    prefs.show_tds_schedule = False
    prefs.epf_withdrawal_taxable = False
    prefs.epf_taxable_percent = 100
    prefs.salary_is_net_of_tds = True
    prefs.updated_at = datetime.now(UTC)

    session.commit()
    session.refresh(prefs)

    return _model_to_response(prefs)


# ----- Section-specific endpoints for granular updates -----


@router.put("/fiscal-year")
def update_fiscal_year(
    current_user: CurrentUser,
    config: FiscalYearConfig,
    session: DatabaseSession,
) -> UserPreferencesResponse:
    """Update fiscal year configuration."""
    return _update_section(session, current_user, config)


@router.put("/essential-categories")
def update_essential_categories(
    current_user: CurrentUser,
    config: EssentialCategoriesConfig,
    session: DatabaseSession,
) -> UserPreferencesResponse:
    """Update essential categories list."""
    return _update_section(session, current_user, config, json_fields={"essential_categories"})


@router.put("/investment-mappings")
def update_investment_mappings(
    current_user: CurrentUser,
    config: InvestmentMappingsConfig,
    session: DatabaseSession,
) -> UserPreferencesResponse:
    """Update investment account mappings."""
    return _update_section(
        session, current_user, config, json_fields={"investment_account_mappings"}
    )


@router.put("/income-sources")
def update_income_sources(
    current_user: CurrentUser,
    config: IncomeSourcesConfig,
    session: DatabaseSession,
) -> UserPreferencesResponse:
    """Update income source category mappings."""
    return _update_section(
        session,
        current_user,
        config,
        json_fields={
            "taxable_income_categories",
            "investment_returns_categories",
            "non_taxable_income_categories",
            "other_income_categories",
        },
    )


@router.put("/capital-loss-categories")
def update_capital_loss_categories(
    current_user: CurrentUser,
    config: CapitalLossConfig,
    session: DatabaseSession,
) -> UserPreferencesResponse:
    """Declare which EXPENSE categories are really realised investment losses.

    This is the ONLY way rows stop counting as spending. Detection never writes
    here on the user's behalf -- it only suggests candidates -- because changing
    the set changes historical expense totals, and that has to be the user's
    call.
    """
    return _update_section(session, current_user, config, json_fields={"capital_loss_categories"})


@router.put("/budget-defaults")
def update_budget_defaults(
    current_user: CurrentUser,
    config: BudgetDefaultsConfig,
    session: DatabaseSession,
) -> UserPreferencesResponse:
    """Update budget default settings."""
    return _update_section(session, current_user, config)


@router.put("/display")
def update_display_preferences(
    current_user: CurrentUser,
    config: DisplayPreferencesConfig,
    session: DatabaseSession,
) -> UserPreferencesResponse:
    """Update display and format preferences."""
    return _update_section(session, current_user, config)


@router.put("/anomaly-settings")
def update_anomaly_settings(
    current_user: CurrentUser,
    config: AnomalySettingsConfig,
    session: DatabaseSession,
) -> UserPreferencesResponse:
    """Update anomaly detection settings."""
    return _update_section(session, current_user, config, json_fields={"anomaly_types_enabled"})


@router.put("/recurring-settings")
def update_recurring_settings(
    current_user: CurrentUser,
    config: RecurringSettingsConfig,
    session: DatabaseSession,
) -> UserPreferencesResponse:
    """Update recurring transaction detection settings."""
    return _update_section(session, current_user, config)


@router.put("/spending-rule")
def update_spending_rule(
    current_user: CurrentUser,
    config: SpendingRuleConfig,
    session: DatabaseSession,
) -> UserPreferencesResponse:
    """Update spending rule target percentages."""
    return _update_section(session, current_user, config)


@router.put("/credit-card-limits")
def update_credit_card_limits(
    current_user: CurrentUser,
    config: CreditCardLimitsConfig,
    session: DatabaseSession,
) -> UserPreferencesResponse:
    """Update credit card limit settings."""
    return _update_section(session, current_user, config, json_fields={"credit_card_limits"})


@router.put("/earning-start-date")
def update_earning_start_date(
    current_user: CurrentUser,
    config: EarningStartDateConfig,
    session: DatabaseSession,
) -> UserPreferencesResponse:
    """Update earning start date configuration."""
    return _update_section(session, current_user, config)


@router.put("/salary-structure")
def update_salary_structure(
    current_user: CurrentUser,
    config: SalaryStructureConfig,
    session: DatabaseSession,
) -> UserPreferencesResponse:
    """Update salary structure for one or more fiscal years."""
    return _update_section(session, current_user, config, json_fields={"salary_structure"})


@router.put("/rsu-grants")
def update_rsu_grants(
    current_user: CurrentUser,
    config: RsuGrantsConfig,
    session: DatabaseSession,
) -> UserPreferencesResponse:
    """Update RSU grants and vesting schedules."""
    return _update_section(session, current_user, config, json_fields={"rsu_grants"})


@router.put("/growth-assumptions")
def update_growth_assumptions(
    current_user: CurrentUser,
    config: GrowthAssumptionsConfig,
    session: DatabaseSession,
) -> UserPreferencesResponse:
    """Update growth assumptions for tax projections."""
    return _update_section(session, current_user, config, json_fields={"growth_assumptions"})


# ----- AI Assistant Configuration -----
